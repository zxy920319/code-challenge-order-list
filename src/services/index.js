import axios from "axios";

const ENDPOINT = "https://api.myjson.com/bins";

const handleError = err => {
  // eslint-disable-next-line no-console
  console.warn(err);
  return false;
};

export async function GetUserInfo() {
  try {
    const res = await axios.get(ENDPOINT + "/pdefl");
    return res.data;
  } catch (err) {
    return handleError(err);
  }
}

export async function GetLastOrders() {
  try {
    const res = await axios.get(ENDPOINT + "/19f9bd");
    return res.data;
  } catch (err) {
    return handleError(err);
  }
}

export async function GetOrderInfo() {
  try {
    const res = await axios.get(ENDPOINT + "/o1sp5");
    return res.data;
  } catch (err) {
    return handleError(err);
  }
}
