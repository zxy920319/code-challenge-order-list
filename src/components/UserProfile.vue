<template>
  <div class="profile-container">
    <v-parallax
      :src="require('@/assets/img/luke-michael-1cWZgnBhZRs-unsplash.jpg')"
      height="400"
    ></v-parallax>

    <v-container class="profile-wrap">
      <v-row align="center" justify="center">
        <v-col class="text-center" cols="12">
          <v-sheet
            v-if="profile"
            elevation="24"
            class="mx-auto"
            height="100vh"
            width="80vw"
          >
            <div class="picture">
              <v-btn fab color="primary" elevation="4">
                <v-avatar size="150">
                  <img
                    :src="profile.picture.large"
                    :alt="`${profile.firstName} ${profile.lastName}`"
                  />
                </v-avatar>
              </v-btn>
            </div>
          </v-sheet>
        </v-col>
      </v-row>
    </v-container>
  </div>
</template>

<script>
import { GetUserInfo } from "@/services";
import { mapState, mapMutations } from "vuex";

export default {
  name: "UserProfile",

  data: () => ({
    //
  }),

  computed: {
    ...mapState(["profile"])
  },

  mounted() {
    if (!this.profile) this.fetchUser();
  },

  methods: {
    ...mapMutations(["setProfile"]),
    async fetchUser() {
      const res = await GetUserInfo();
      // eslint-disable-next-line no-console
      console.log(res);

      if (res) this.setProfile(res);
    }
  }
};
</script>

<style lang="scss">
.profile-container {
  height: calc(100vh + 400px);
  background-color: var(--v-accent-lighten2);
  .profile-wrap {
    position: absolute;
    top: 300px;
  }
  .picture {
    margin-top: -64px;
    display: inline-block;
    button {
      width: 160px;
      height: 160px;
    }
  }
}
</style>
