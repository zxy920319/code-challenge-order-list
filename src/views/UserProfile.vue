<template>
  <div class="profile-container">
    <v-parallax
      :src="require('@/assets/img/luke-michael-1cWZgnBhZRs-unsplash.jpg')"
      height="400"
    ></v-parallax>

    <v-container class="profile-wrap">
      <v-row align="center" justify="center">
        <v-col class="text-center" cols="12">
          <v-sheet v-if="profile" elevation="24" class="mx-auto" width="90vw">
            <v-row justify="center" align="center">
              <v-col cols="12" md="4">
                <div class="picture">
                  <v-btn fab color="primary" elevation="4">
                    <v-avatar size="150">
                      <img :src="profile.picture.large" :alt="name" />
                    </v-avatar>
                  </v-btn>
                </div>
                <v-col
                  cols="12"
                  md="12"
                  class="d-flex justify-center align-center  headline accent--text text--darken-4"
                >
                  {{ name }}
                  <v-icon v-if="profile.gender === 'Female'" color="pink"
                    >mdi-gender-female</v-icon
                  >
                  <v-icon v-else color="blue">mdi-gender-male</v-icon>
                </v-col>
                <div
                  class="d-flex justify-center align-center mt-6 headline accent--text text--darken-4"
                >
                  <v-icon small color="primary" class="mr-4">mdi-email</v-icon>
                  <div class="body-1 mt-1">{{ profile.emailAddress }}</div>
                </div>
                <div
                  class="d-flex justify-center align-center mt-4 headline accent--text text--darken-4"
                >
                  <v-icon small color="primary" class="mr-4">mdi-cake</v-icon>
                  <div class="body-1 mt-1">{{ dob }}</div>
                </div>
                <div
                  class="d-flex justify-center align-center mt-4 headline accent--text text--darken-4"
                >
                  <v-icon small color="primary" class="mr-4">mdi-home</v-icon>
                  <div class="body-1 mt-1">{{ profile.address }}</div>
                </div>
              </v-col>
            </v-row>

            <router-view />
          </v-sheet>
        </v-col>
      </v-row>
    </v-container>
  </div>
</template>

<script>
import { GetUserInfo } from "@/services";
import { mapState, mapMutations } from "vuex";
import * as moment from "moment";
export default {
  name: "UserProfile",

  computed: {
    ...mapState(["profile"]),
    name() {
      return `${this.profile.firstName} ${this.profile.lastName}`;
    },
    dob() {
      return moment(this.profile.birthday).format("D MMM YYYY");
    }
  },

  mounted() {
    if (!this.profile) this.fetchUser();
  },

  methods: {
    ...mapMutations(["setProfile"]),
    async fetchUser() {
      const res = await GetUserInfo();
      // eslint-disable-next-line no-console
      console.log(res);

      if (res) this.setProfile(res);
    }
  }
};
</script>
