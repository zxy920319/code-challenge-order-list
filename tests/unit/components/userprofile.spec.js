import Vuetify from "vuetify";
import Vuex from "vuex";
import { shallowMount, createLocalVue } from "@vue/test-utils";
import UserProfile from "@/views/UserProfile.vue";
import { GetUserInfo } from "@/services";

jest.mock("@/services");
GetUserInfo.mockResolvedValue({});

const localVue = createLocalVue();

localVue.use(Vuex);
localVue.use(Vuetify);

let mutation = {
  setProfile: state => (state.profile = {})
};

let store = new Vuex.Store({
  state: {
    profile: {
      customerId: 8481133476,
      firstName: "John",
      lastName: "Doe",
      gender: "Male",
      language: "en",
      currency: "euro",
      emailAddress: "john.doe@example.com",
      birthday: "1988-07-02",
      address: "8923 Mockingbird Hill",
      registered: { date: "2015-11-04T22:09:36Z", age: 3 },
      picture: {
        large: "https://randomuser.me/api/portraits/men/81.jpg",
        medium: "https://randomuser.me/api/portraits/med/men/81.jpg",
        thumbnail: "https://randomuser.me/api/portraits/thumb/men/81.jpg"
      }
    }
  },
  mutation
});

let vuetify = new Vuetify();

jest.mock("moment", () => () => ({
  format: () => "2018–01–30T12:34:56+00:00"
}));

describe("UserProfile", () => {
  it("should render class profile-container", () => {
    const wrapper = shallowMount(UserProfile, {
      store,
      vuetify,
      localVue,
      computed: {
        dob() {
          return "date";
        }
      }
    });
    expect(wrapper.classes()).toContain("profile-container");
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("GetUserInfo will not be called if profile exists in central state", () => {
    expect(GetUserInfo.mock.calls.length).toBe(0);
  });
});
