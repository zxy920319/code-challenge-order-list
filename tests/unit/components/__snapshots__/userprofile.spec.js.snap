// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserProfile should render class profile-container 1`] = `
<div class="profile-container">
  <v-parallax-stub height="400" alt="" src=""></v-parallax-stub>
  <v-container-stub tag="div" class="profile-wrap">
    <v-row-stub tag="div" align="center" justify="center">
      <v-col-stub cols="12" tag="div" class="text-center">
        <v-sheet-stub elevation="24" minheight="100vh" width="90vw" tag="div" class="mx-auto">
          <v-row-stub tag="div" align="center" justify="center">
            <v-col-stub cols="12" md="4" tag="div">
              <div class="picture">
                <v-btn-stub color="primary" elevation="4" tag="button" activeclass="" fab="true" type="button">
                  <v-avatar-stub size="150"><img src="https://randomuser.me/api/portraits/men/81.jpg" alt="John Doe"></v-avatar-stub>
                </v-btn-stub>
              </div>
              <v-col-stub cols="12" md="12" tag="div" class="d-flex justify-center align-center  headline accent--text text--darken-4">
                John Doe
                <v-icon-stub color="blue">mdi-gender-male</v-icon-stub>
              </v-col-stub>
              <div class="d-flex justify-center align-center mt-6 headline accent--text text--darken-4">
                <v-icon-stub small="" color="primary" class="mr-4">mdi-email</v-icon-stub>
                <div class="body-1 mt-1">john.doe@example.com</div>
              </div>
              <div class="d-flex justify-center align-center mt-4 headline accent--text text--darken-4">
                <v-icon-stub small="" color="primary" class="mr-4">mdi-cake</v-icon-stub>
                <div class="body-1 mt-1">date</div>
              </div>
              <div class="d-flex justify-center align-center mt-4 headline accent--text text--darken-4">
                <v-icon-stub small="" color="primary" class="mr-4">mdi-home</v-icon-stub>
                <div class="body-1 mt-1">8923 Mockingbird Hill</div>
              </div>
            </v-col-stub>
          </v-row-stub>
          <router-view></router-view>
        </v-sheet-stub>
      </v-col-stub>
    </v-row-stub>
  </v-container-stub>
</div>
`;
